### PR TITLE
fix(es/minifier): Don't track usage assign

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**
Don't know if it's right, but it generate correct output.

**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

File: crates/swc_ecma_minifier/src/analyzer/storage/normal.rs
Comment (kdy1) [inline crates/swc_ecma_minifier/src/analyzer/storage/normal.rs:276]: Please remove this line

File: crates/swc_ecma_minifier/src/analyzer/storage/normal.rs
Comment (kdy1) [inline crates/swc_ecma_minifier/src/analyzer/storage/normal.rs:279]: Same here

File: crates/swc_ecma_minifier/src/analyzer/storage/normal.rs
Comment (kdy1) [inline crates/swc_ecma_minifier/src/analyzer/storage/normal.rs:281]: Same here

File: crates/swc_ecma_minifier/src/analyzer/storage/normal.rs
Comment (kdy1) [inline crates/swc_ecma_minifier/src/analyzer/storage/normal.rs:290]: Same here

File: crates/swc_ecma_minifier/src/analyzer/storage/normal.rs
Comment (kdy1) [inline crates/swc_ecma_minifier/src/analyzer/storage/normal.rs:260]: Same here

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thank you

**Comment by Austaras** [comment]
Oh, sorry

File: crates/swc_ecma_minifier/tests/benches-full/lodash.js
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: I think this can be problematic if `pair[0]` modifies `toString`

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: It's the goal of infection analysis (although I'm not sure if this is a correct name)

File: crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js
Comment (kdy1) [inline crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js:2760]: Same here. We need to prove that `objGetter(self)` does not modify `isObject`.

**Review by kdy1 — COMMENTED** [review]
I'm not sure if we are proving these facts.
I added two-way infection, which is clearly not optimal, because it prevents bugs

File: crates/swc_ecma_minifier/tests/benches-full/lodash.js
Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: I've tested this code in terser and it generates identical output.

> I think this can be problematic if pair[0] modifies toString

Do you mean access to pair[0] may modify toString?

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Yes, we may need more execution tests with getters with side effects. e.g.

```js

let foo = () => 1;

const obj = {
    get 0() {
        foo = () => 2
        return 40
    }
};

```

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Oh, I have an idea. I'll review this by diffing with terser

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Seems like this to be the same reason as terser/terser/#1244

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Oh... I see. Thank you!

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: In this case
```js
let foo = ()=>1;
const obj = {
    get 0 () {
        return foo = ()=>2, 40;
    }
};
var c = obj(0);
console.log(foo(c));
```
this PR would generate
```js
let foo = ()=>1;
const obj = {
    get 0 () {
        return foo = ()=>2, 40;
    }
};
var c = obj(0);
console.log(foo(c));
```
So I guess it's fine.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: It's not the exact problematic case. We can't know input.
Sometimes there are even already-minified code in the input, which is hard to imagine

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Oh, wait, this is wrong, let me recheck it.

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: So, for code
```js
let foo = () => 1;

const obj = {
    get 0() {
        foo = () => 2;
        return 40;
    },
};
console.log(obj)

var c = obj[0];

console.log(foo(c));
```
SWC would incorrectly inline c because of #5961. I believe the solution is very simple -- any reassignment to foo should prevent seq inline. However this has nothing to do with this PR.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Yeah ignoring side effects of getters is fine, as terser does it too

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Problem is the case below

Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: No, terser wouldn't inline `c` here.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Oh, so infection analysis is not catching the related of `obj` and `foo`.
Maybe I need to file an issue...

File: crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js
Comment (Austaras) [reply crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js:2760]: Same here. terser would inline newValue as long ass isObject is declared.

File: crates/swc_ecma_minifier/tests/benches-full/lodash.js
Comment (Austaras) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: I don't think it has any relation with infection analysis in this case. Any reassignment to `foo` should prevent seq inline.

Comment (kdy1) [reply crates/swc_ecma_minifier/tests/benches-full/lodash.js:1507]: Oh yes it would be better

File: crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js
Comment (kdy1) [reply crates/swc_ecma_minifier/tests/projects/output/angular-1.2.5.js:2760]: No, the declaration is not the condition. I changed the definition of `isObject` to `var isObject = globalThis.isObject;` and terser preserves 

```js
newValue = objGetter(self);
                        var newLength, key;
                        if (!isObject(newValue)) {
                            if (oldValue !== newValue) {
                                oldValue = newValue;
                                changeDetected++
                            }
```

https://gist.github.com/kdy1/c4247944514115421aaed0c9288b6d38#file-beutified-js-L5291-L5293

https://gist.github.com/kdy1/c4247944514115421aaed0c9288b6d38

**Review by kdy1 — CHANGES_REQUESTED** [review]

**Comment by Austaras** [comment]
Yeah, if `isObject` is a normal function it would work fine. Maybe there's some more complex mechanism.

**Comment by Austaras** [comment]
The difference can be run down to https://github.com/terser/terser/blob/master/lib/compress/tighten-body.js#L887. It checks if it's a AST_SymbolDefun(definition of function) or defined in current scope, which I believe is just a convenient implementation.

**Comment by Austaras** [comment]
> The difference can be run down to https://github.com/terser/terser/blob/master/lib/compress/tighten-body.js#L887. It checks if it's a AST_SymbolDefun(definition of function) or defined in current scope, which I believe is just a convenient implementation.

So basically `function isObject` is fine, but `const isObject = function` is fine only if it's defined in same scope of `newValue`.

**Review by jridgewell — APPROVED** [review]

**Review by kdy1 — APPROVED** [review]
This PR seems innocent.
I'll merge this after merging https://github.com/swc-project/swc/pull/6053

---


swc-bump:
 - swc_ecma_minifier

**Review by kdy1 — APPROVED** [review]
Thank you!

**Pull request merged**
